src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestState.cs

```diff
@@ -123,6 +123,22 @@ public void ClearSendRequestState()
         public long? ExpectedBytesToRead { get; set; }
         public long CurrentBytesRead { get; set; }
 
+        // TODO (Issue 2505): temporary pinned buffer caches of 1 item. Will be replaced by PinnableBufferCache.
+        private GCHandle _cachedReceivePinnedBuffer = default(GCHandle);
+
+        public void PinReceiveBuffer(byte[] buffer)
+        {
+            if (!_cachedReceivePinnedBuffer.IsAllocated || _cachedReceivePinnedBuffer.Target != buffer)
```

Comment:
What guarantees do we have that there isn't still an outstanding operation using the old target?  Maybe there's something we can assert, about there not being any outstanding operations or something?